idaes/core/util/model_diagnostics.py

```diff
@@ -0,0 +1,577 @@
+''' Degeneracy Hunter is a collection of utility functions to assist in mathematical
+modeling in Pyomo.
+
+Alex Dowling, University of Notre Dame
+
+'''
+
+from pyomo.environ import *
+from pyomo.core.kernel.component_set import ComponentSet
+from pyomo.core.expr.visitor import identify_variables
+from pyomo.contrib.pynumero.interfaces.pyomo_nlp import PyomoNLP
+import numpy as np
+from scipy.sparse.linalg import svds
+from scipy.sparse import issparse, find
+
+from pyomo.opt import SolverStatus, TerminationCondition
+
+import matplotlib.pyplot as plt
+
+
+# This library is already used in Pyomo
+import operator
+
+'''
+This code is from Robby Parker.
+'''
+
+def get_con_eq_idx_map(interface, constraint_names):
+    # This is super janky. TODO: Add functionality
+    # to PyNumero.
+    # constraint_names is an argument because generating
+    # constraint_names list from PyNumero takes forever.
+    # TODO: Find way to make AslNLP::constraint_names()
+    # not ridiculously slow.
+    mask = interface._con_full_eq_mask
+    temp = np.array(constraint_names)
+    eq_names = list(temp[mask])
+    return {name: i for i, name in enumerate(eq_names)}
```

Comment:
I believe a better implementation would be

```python
def get_con_eq_idx_map(nlp):
    eq_cons = get_eq_con_list(nlp) # See my implementation of function below
    return ComponentMap((con, i) for i, con in enumerate(eq_cons))
```
This avoids needing to generate constraint names, which is slow.